File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriverFull.cs
#region test_driver_1
using Raven.Client.Documents;
using Raven.TestDriver;
using Xunit;
using System;
using System.IO;
using System.Linq;
using Raven.Client.Documents.Indexes;

namespace RavenDBTestDriverFullExample
{
    public class RavenDBTestDriver : RavenTestDriver<MyRavenDBLocator>
    {
        //This allows us to modify the conventions of the store we get from 'GetDocumentStore'
        protected override void PreInitialize(IDocumentStore documentStore)
        {
            documentStore.Conventions.MaxNumberOfRequestsPerSession = 50;
        }

        [Fact]
        public void MyFirstTest()
        {
            using (var store = GetDocumentStore())
            {
                store.ExecuteIndex(new TestDocumentByName());
                using (var session = store.OpenSession())
                {
                    session.Store(new TestDocument { Name = "Hello world!" });
                    session.Store(new TestDocument { Name = "Goodbye..." });
                    session.SaveChanges();
                }
                WaitForIndexing(store); //If we want to query documents sometime we need to wait for the indexes to catch up
                WaitForUserToContinueTheTest(store);//Sometimes we want to debug the test itself, this redirect us to the studio
                using (var session = store.OpenSession())
                {
                    var query = session.Query<TestDocument, TestDocumentByName>().Where(x => x.Name == "hello").ToList();
                    Assert.Single(query);
                }
            }
        }
    }

    public class MyRavenDBLocator : RavenServerLocator
    {
        private string _serverPath;
        private string _command = "dotnet";
        private readonly string RavenServerName = "Raven.Server";
        private string _arguments;

        public override string ServerPath
        {
            get
            {
                if (string.IsNullOrEmpty(_serverPath) == false)
                {
                    return _serverPath;
                }
                var path = Environment.GetEnvironmentVariable("RavenServerTestPath");
                if (string.IsNullOrEmpty(path) == false)
                {
                    if (InitializeFromPath(path))
                        return _serverPath;
                }
                //If we got here we didn't have ENV:RavenServerTestPath setup for us maybe this is a CI environment
                path = Environment.GetEnvironmentVariable("RavenServerCIPath");
                if (string.IsNullOrEmpty(path) == false)
                {
                    if (InitializeFromPath(path))
                        return _serverPath;
                }
                //We couldn't find Raven.Server in either environment variables let's look for it in the current directory
                foreach (var file in Directory.GetFiles(Environment.CurrentDirectory, $"{RavenServerName}.exe; {RavenServerName}.dll"))
                {
                    if (InitializeFromPath(file))
                        return _serverPath;
                }
                //Lets try some brut force
                foreach (var file in Directory.GetFiles(Directory.GetDirectoryRoot(Environment.CurrentDirectory), $"{RavenServerName}.exe; {RavenServerName}.dll", SearchOption.AllDirectories))
                {
                    if (InitializeFromPath(file))
                    {
                        try
                        {
                            //We don't want to override the variable if defined
                            if (string.IsNullOrEmpty(Environment.GetEnvironmentVariable("RavenServerTestPath")))
                                Environment.SetEnvironmentVariable("RavenServerTestPath", file);
                        }
                        //We might not have permissions to set the enviroment variable
                        catch
                        {

                        }
                        return _serverPath;
                    }
                }
                throw new FileNotFoundException($"Could not find {RavenServerName} anywhere on the device.");
            }
        }

        private bool InitializeFromPath(string path)
        {
            if (Path.GetFileNameWithoutExtension(path) != RavenServerName)
                return false;
            var ext = Path.GetExtension(path);
            if (ext == ".dll")
            {
                _serverPath = path;
                _arguments = _serverPath;
                return true;
            }
            if (ext == ".exe")
            {
                _serverPath = path;
                _command = _serverPath;
                _arguments = string.Empty;
                return true;
            }
            return false;
        }

        public override string Command => _command;
        public override string CommandArguments => _arguments;
    }

    public class TestDocumentByName : AbstractIndexCreationTask<TestDocument>
    {
        public TestDocumentByName()
        {
            Map = docs => from doc in docs select new { doc.Name };
            Indexes.Add(x => x.Name, FieldIndexing.Search);
        }
    }

    public class TestDocument
    {
        public string Name { get; set; }
    }
}
#endregion

File: Documentation/4.0/Samples/java/src/test/java/net/ravendb/start/RavenDBTestDriverFull.java
package net.ravendb.start;

import net.ravendb.client.documents.DocumentStore;
import net.ravendb.client.documents.IDocumentStore;
import net.ravendb.client.documents.indexes.AbstractIndexCreationTask;
import net.ravendb.client.documents.indexes.FieldIndexing;
import net.ravendb.client.documents.session.IDocumentSession;
import net.ravendb.client.test.driver.RavenServerLocator;
import net.ravendb.client.test.driver.RavenTestDriver;
import org.junit.Assert;
import org.junit.Test;

import java.util.List;

public class RavenDBTestDriverFull {
    //region test_driver_1

    public class MyRavenDBTestDriver extends RavenTestDriver {
        public MyRavenDBTestDriver(RavenServerLocator locator, RavenServerLocator securedLocator) {
            super(new MyRavenDBLocator(), null);
        }

        //This allows us to modify the conventions of the store we get from 'getDocumentStore'
        @Override
        protected void customizeStore(DocumentStore store) {
            store.getConventions().setMaxNumberOfRequestsPerSession(50);
        }

        @Test
        public void myFirstTest() throws Exception {
            try (IDocumentStore store = getDocumentStore()) {
                store.executeIndex(new RavenDBTestDriver.TestDocumentByName());

                try (IDocumentSession session = store.openSession()) {
                    TestDocument document1 = new TestDocument();
                    document1.setName("Hello world!");

                    TestDocument document2 = new TestDocument();
                    document2.setName("Goodbye...");

                    session.store(document1);
                    session.store(document2);
                    session.saveChanges();
                }

                waitForIndexing(store);  //If we want to query documents sometime we need to wait for the indexes to catch up
                waitForUserToContinueTheTest(store); //Sometimes we want to debug the test itself, this redirect us to the studio

                try (IDocumentSession session = store.openSession()) {
                    List<RavenDBTestDriver.TestDocument> query = session
                            .query(
                                RavenDBTestDriver.TestDocument.class,
                                RavenDBTestDriver.TestDocumentByName.class
                            )
                        .whereEquals("name", "hello")
                        .toList();

                    Assert.assertEquals(1, query.size());
                }
            }
        }
    }

    public class MyRavenDBLocator extends RavenServerLocator {
        @Override
        public String getServerPath() {
            return "/opt/RavenDB/Server/Raven.Server";
        }

        @Override
        public String getCommand() {
            return getServerPath();
        }
    }

    public static class TestDocumentByName extends AbstractIndexCreationTask {
        public TestDocumentByName() {
            map = "from doc in docs select new { doc.name }";
            index("name", FieldIndexing.SEARCH);
        }
    }

    public class TestDocument {
        private String name;

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }
    }
    //endregion
}
